reco_utils/evaluation/spark_evaluation.py

```diff
@@ -470,3 +479,457 @@ def _get_relevant_items_by_timestamp(
+                    ),
+                    on="i2",
+                )
+                .select(
+                    "i1",
+                    "i2",
+                    (
+                        F.col("count")
+                        / (F.col("i1_sqrt_count") * F.col("i2_sqrt_count"))
+                    ).alias(self.sim_col),
+                )
+                .repartition(n_partitions, "i1", "i2")
+            )
+        return self.df_cosine_similarity
+
+    # Diversity metrics
+    def _get_intralist_similarity(self, df):
+        """Intra-list similarity from
+
+        :Citation:
+
+            "Improving Recommendation Lists Through Topic Diversification",
+            Ziegler, McNee, Konstan and Lausen, 2005.
+        """
+        if self.df_intralist_similarity is None:
+            pairs = self._get_pairwise_items(df=df)
+            similarity_df = self._get_cosine_similarity()
+            # Fillna(0) is needed in the cases where similarity_df does not have an entry for a pair of items.
+            # e.g. i1 and i2 have never occurred together.
+            self.df_intralist_similarity = (
+                pairs.join(similarity_df, on=["i1", "i2"], how="left")
+                .fillna(0)
+                .filter(F.col("i1") != F.col("i2"))
+                .groupBy(self.col_user)
+                .agg(F.mean(self.sim_col).alias("avg_il_sim"))
+                .select(self.col_user, "avg_il_sim")
+            )
+        return self.df_intralist_similarity
+
+    def user_diversity(self):
+        """Calculate average diversity of recommendations for each user.
+        The metric definition is based on formula (3) in the following reference:
+
+        :Citation:
+
+            Y.C. Zhang, D.Ó. Séaghdha, D. Quercia and T. Jambor, Auralist:
+            introducing serendipity into music recommendation, WSDM 2012
+
+        Returns:
+            pyspark.sql.dataframe.DataFrame: A dataframe with the following columns: col_user, user_diversity.
+        """
+        if self.df_user_diversity is None:
+            self.df_intralist_similarity = self._get_intralist_similarity(self.reco_df)
+            self.df_user_diversity = (
+                self.df_intralist_similarity.withColumn(
+                    "user_diversity", 1 - F.col("avg_il_sim")
+                )
+                .select(self.col_user, "user_diversity")
+                .orderBy(self.col_user)
+            )
+        return self.df_user_diversity
+
+    def diversity(self):
+        """Calculate average diversity of recommendations across all users.
+
+        Returns:
+            float: diversity.
+        """
+        if self.df_diversity is None:
+            self.df_user_diversity = self.user_diversity()
+            self.df_diversity = self.df_user_diversity.agg(
+                {"user_diversity": "mean"}
+            ).first()[0]
+        return self.df_diversity
+
+    # Novelty metrics
+    def item_novelty(self):
```

Comment:
Now I have removed user novelty, equation (7) is not used anywhere in the code.

The definition of item novelty you had came from the blog post above. I changed this definition to conform with the Castells et al. paper. I suggest we stick to the paper instead of a blog post for several reasons, including that papers in the literature follow the same approach which is based on a probabilistic model. Moreover, the papers have been peer reviewed whereas a blog post reflects just one person's ideas. 

What happened was that the blogger changed the standard definition because he misunderstands these models in the literature. He proposes his own definition but it is incorrect and doesn't correspond to any actual probability. It is basically justified on rough intuition.